docs/notebooks/openai_gym_lunar_lander.pct.py

```diff
@@ -0,0 +1,253 @@
+
+# %%
+for _ in range(10):
+    sample_w = search_space.sample(1).numpy()[0]
+    demo_heuristic_lander(env, sample_w, print_reward=True)
+
+# %% [markdown]
+# As you can see, most of the random sets of parameters result in a negative reward. So picking a value from this search space at random can result in a various unwanted behaviors. Here we show some examples of the landing not going according to plan. Each of these examples was created with a sample of the parameter values from the search space.
+#
+# **Warning:** all the videos in this notebook were pre-generated. Creating renders of OpenAI Gym environments requires various dependencies depending on software setups and operating systems, so we have chosen not to do it here in the interest of transferability of this notebook. For those interested in reproducing these videos, we have saved the input parameters and the code we used to generate them in the Trieste repository, in the folder next to this notebook. However because of the stochastic nature of the environment and the optimization described here, you results might differ slightly from those shown here.
+
+# %%
+import io
+import base64
+from IPython.display import HTML
+
+def load_video(filename):
+    video = io.open('./lunar_lander_videos/' + filename, 'r+b').read()
+    encoded = base64.b64encode(video)
+    return HTML(data='''
+            <video width="360" height="auto" alt="test" controls>
+                <source src="data:video/mp4;base64,{0}" type="video/mp4" />
+            </video>'''.format(encoded.decode('ascii'))
+    )
+
+
+# %% [markdown]
+# #### Crash
+
+# %%
+load_video("crash.mp4")
+
+# %% [markdown]
+# #### Timeout
+
+# %%
+load_video("timeout.mp4")
+
+# %% [markdown]
+# #### Flying out of bounds
+
+# %%
+load_video("out_of_bounds.mp4")
+
+# %% [markdown]
+# #### Slamming on the surface
+# This is a very common failure mode in this environment - going too fast and slamming on the surface.
+
+# %%
+load_video("slam.mp4")
+
+# %% [markdown]
+# Finally, we need a way to deal with stochasticity of the simulation. To keep our modelling simple, let's aim at finding a controller that is good on average over a certain number of runs. This gives us a simple way of implementing the Trieste observer. Also note that we reverse the reward sign: this is because we seek the maximum reward, and Trieste approaches optimization as minimization problem.
+
+# %%
+N_RUNS = 10
+def lander_objective(x):
+    # for each point compute average reward over n_runs runs
+    all_rewards = []
+    for w in x.numpy():
+        rewards = [
+            demo_heuristic_lander(env, w)
+            for _ in range(N_RUNS)
+        ]
+        all_rewards.append(rewards)
+
+    rewards_tensor = tf.convert_to_tensor(all_rewards, dtype=tf.float64)
+    
+    # triste minimizes, and we want to maximize
+    return -1 * tf.reshape(tf.math.reduce_mean(rewards_tensor, axis=1), (-1, 1))
```

Comment:
Not sure if worth mentioning here, but we could /should actually use the variance of the rewards_tensor to estimate the likelihood variance, since the noise is likely to be heteroscedastic.
Turns out it's not really easy to do this in GPflow, you need a custom VGP model as described here: 
https://gpflow.readthedocs.io/en/master/notebooks/advanced/varying_noise.html
but that would be the "proper" way to solve this problem.

Reply:
Shall we leave a comment here? Or do you reckon we shall actually do it this way?

Reply:
Either way is fine. The VGP model should bring a bit of an extra performance, but it's probably 50 lines of extra code, so just leaving a comment that we could do better could suffice.

Reply:
A comment is left in the `build_model` function